moon/controller_round3.py

```diff
@@ -447,38 +410,81 @@ def on_scan(self, timestamp, data):
-                        self.homebase_final_approach_distance = float("inf")
-                        self.object_reached('homebase')
-                    else:
-                        # keep going straight; for now this means do nothing, keep speed from previous step
-                        self.currently_following_object['timestamp'] = self.time # freshen up timer as we are still following homebase
+                if straight_ahead_dist < HOMEBASE_KEEP_DISTANCE:
+                    self.publish("desired_movement", [0, 0, 0])
 
-            # we expect that lidar bounces off of homebase as if it was a big cylinder, not taking into consideration legs, etc.
+                    print ("app: final homebase distance %f: " % straight_ahead_dist)
+                    self.object_reached('homebase')
 
             if 'basemarker' in self.objects_to_follow:
 
+                # find min_index where distance < 5 and max_index where distance < 5
+                # find circle passing through these points (defined in polar coordinates as [index*6 degrees, distance]
+                min_index = max_index = -1
+                for i in range(0,180):
+                    if data[i]> 10 and data[i] < 5000:
+                        min_index = i
+                        break
+                for i in range(180, 0, -1):
+                    if data[i-1]> 10 and data[i-1] < 5000:
+                        max_index = i
+                        break
+
+                if min_index < 0 or max_index < 0:
+                    return
+                
+                def pol2cart(rho, phi):
+                    x = rho * math.cos(phi)
+                    y = rho * math.sin(phi)
+                    return(x, y)
+
+                # Best Fit Circle https://goodcalculators.com/best-fit-circle-least-squares-calculator/
+                x_l=[]
+                y_l=[]
+                for i in range(min_index, max_index):
+                    x,y = pol2cart(data[i] / 1000.0, -1.29999995232 + (i - 40) * 0.0262626260519)
+                    x_l.append(x)
+                    y_l.append(y)
+
+                nop = len(x_l)       
+                x = np.array(x_l)
+                y = np.array(y_l)
+
+                x_y = np.multiply(x,y)
+                x_2 = np.square(x)
+                y_2 = np.square(y)
+
+                x_2_plus_y_2 = np.add(x_2,y_2)
+                x__x_2_plus_y_2 = np.multiply(x,x_2_plus_y_2)
+                y__x_2_plus_y_2 = np.multiply(y,x_2_plus_y_2)
+
+                sum_x = x.sum(dtype=float)
+                sum_y = y.sum(dtype=float)
+                sum_x_2 = x_2.sum(dtype=float)
+                sum_y_2 = y_2.sum(dtype=float)
+                sum_x_y = x_y.sum(dtype=float)
+                sum_x_2_plus_y_2 = x_2_plus_y_2.sum(dtype=float)
+                sum_x__x_2_plus_y_2 = x__x_2_plus_y_2.sum(dtype=float)
+                sum_y__x_2_plus_y_2 = y__x_2_plus_y_2.sum(dtype=float)
```

Comment:
nice, I do not believe you -> I would move this to separate function and add some some tests that it is doing what is expected

Reply:
unit test has been added, generates 100 random circles, reports scan data with 5% added error, checks extrapolated circle against the true one

Reply:
:) this is then overkill on the other side ;). Are the random circles with fixed seed, i.e. if it once fails it will always fail? Maybe 3 would be sufficient?? I will have a look, thanks.

Reply:
`python -m unittest moon\test_controller_round3.py` is blocking (displays random circles and rays) - that is not desired as tests are running automatically for several setups (OS and Python version)

Reply:
The idea was to make graphical representation available so that it is obvious that the test is running meaningful scenarios. The initial motivation for this request for a unit test was that the code seemed to be a black box that can't be validated just by looking at it. Well, if we add another black box of code that interacts with the original one, it is not helping anything, hence the associated graphics. I understand this is not applicable for automated testing, that's why the graphics only kicks in when a graphics package is available on the test platform. 

However, it can certainly happen that a platform intended for testing has the graphics package installed. So my question is - how would you offer enabling of this graphical feature? Some kind on an option or a switch or detecting some other context that would determine whether to show graphics or not?

Reply:
I would only keep the lines commented out (ideally just one line calling "draw circle" ... and I would maybe even move the imports of matplotlib there) ... and move on. Thanks m.